### ISSUE TITLE
Support T9

### ISSUE THREAD

**Issue by maverick74** [florisboard/florisboard#165]
Would be great to have T9 keyboard layout and prediction support.

writing in a qwerty keyboard with big fingers is terrible. This keeps stuck on TouchPal...

Any chance for this?

Thanks

**Comment by patrickgold** [comment]
Hmm, definitely interesting to implement, although I am currently not sure if it will find its way into FlorisBoard, as this is a more difficult thing to implement correctly. Will place it on the maybe list for v0.5.0 though!

**Comment by maverick74** [comment]
If you make a decision on this, please say something over here!

If this goes forward I'll translate Floris into Portuguese ;)

**Comment by patrickgold** [comment]
I will definitely let you know if I end up implementing the T9 layout into FlorisBoard or not. If I end up doing so, I want it to be fully implemented, not a half-working thing. The decision if it will find its way into FlorisBoard will be done when the development for the v0.5.0 milestone begins, which I don't expect before the end of February.

**Comment by maverick74** [comment]
Well, if it is of any use I'm not the only one asking for this, there are a bunch of other people asking for this :)

But I liked a lot that, if you decide to implement it, it will be a complete implementation, not something unfinished

**Comment by maverick74** [comment]
Hi @patrickgold 

anything worth sharing about the decision over T9? :)

**Comment by Waelwindows** [comment]
@maverick74 i think it was clear that the decision will take place when v0.5.0 starts development.

Florisboard is still working on it's v0.4.0 milestone.

**Comment by patrickgold** [comment]
@maverick74 The only thing that's more clear is that quite a lot users want T9, so there's a very high chance it will come sometime when the v0.5.0 milestone development starts. Currently I am still working on the v0.4.0 milestone and T9 is not planned for this one.

**Comment by maverick74** [comment]
@Waelwindows
yes, it was clear! But I kept the end of February in mind (because development on 0.5 was not expected to start before that) and I assumed - wrongly - that it would be at least in 0.4 already... 

Hence my comment. 

Only after I checked the version, but opted not to change the comment as it was already sent! - despite the noise, I believe no one got hurt :)

**Comment by GitHub-user-2020** [comment]
### Discussed in https://github.com/florisboard/florisboard/discussions/1367

<div type='discussions-op-text'>

<sup>Originally posted by **GitHub-user-2020** October 30, 2021</sup>
Namely something like [this](https://ytprivate.com/watch?v=Su4-_v2qGvQ).

I'm aware of only one open source project that implements this (AnySoftKeyboard) but that project seems inactive and their implementation of T9 requires more work: prediction is not based on the key pressed, but rather, on what letter has been returned after the number of presses to return a specific letter as input.

For instance, typing the numbers with T9 prediction "269" can return either the numbers themselves, or with an English language keypad layout and dictionary the words "any", "box" or "boy". With AnySoftKeyboard's implementation, the user would not be able to type the words "box" or "boy" without pressing the number 2 twice (in order to type the letter "b" first before returning suggestions).

The current T9 keyboard app I use is Multiling O Keyboard, which thankfully does not have AnySoftKeyboard's hasty T9 implementation. However, the main issues with Multiling O Keyboard are that it's not open source (to my knowledge) and it is almost exclusively distributed through the Google Play Store (which leaves me and/or others with few, if any de-Googled or even general alternatives to a keyboard which I believe is best suited for small devices).

Thank you in advance for taking the time to consider my suggestion. I would very much appreciate it if you were able to implement this as soon as it's possible or provide an explanation and/or possible recommendations if you can't implement this at all.</div>

**Comment by STPKITT** [comment]
AnySoftKeyboard isn't inactive but as far as I can see from my install (it's the only keyboard I use on Android since years) it doesn't offer a T9 keyboard for general usage. A T9 type layout only shows up in situations in which you click on an input field which only accepts numbers being typed in.

**Comment by GitHub-user-2020** [comment]
> AnySoftKeyboard isn't inactive but as far as I can see from my install (it's the only keyboard I use on Android since years) it doesn't offer a T9 keyboard for general usage. A T9 type layout only shows up in situations in which you click on an input field which only accepts numbers being typed in.
    
Regarding this, AnySoftKeyboard has a 16 key layout that, to me, resembles T9 or the keypad from phones in the 2000s, but it doesn't work as well as I would like it to.
    
[Linked here](https://streamable.com/dvictd) is a video demonstrating the keyboard that I currently use (Multiling O Keyboard, presumably closed-source/proprietary and distributed almost exclusively through the Google Play Store) and AnySoftKeyboard.

**Comment by maverick74** [comment]
I use TouchPal and their T9 is really amazing.

the thing is, as they are closed source, I don't know how much I can trust them.

And that's the whole thing: I want a nice keyboard, open-source that has T9.

ASK is out of the race on this because they don't seem to have any interest in T9.
That leaves Floris as the only alternative...

**Comment by jonnyledger** [comment]
Love this project, t9 and other compact layouts ( eg compact  qwerty) would be a welcome addition for those of us with small phones and big fingers

**Comment by maverick74** [comment]
@patrickgold any new milestone where this is planned?

**Comment by gylas** [comment]
+1

**Comment by khlsvr** [comment]
There is tt9 https://github.com/sspanak/tt9 but they are lacking clipboard history / pinning and also they use their own dictionary format and I am relying heavily on word conversions (for example I type `bulbasaur` and I get `フシギダネ` as a suggestion) that work well with android's built-in dictionaries.
